### PR TITLE
rg-146: Add stripe event dispatcher

### PR DESCRIPTION
**Pull request by jurok3x**
* Introduced stripe event dispatcher.
* Introduced stripe event handlers.

### CODE REVIEW

File: backend/src/bundles/stripe-events/handlers/subscription-created-event-handler.ts [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+        const subscription: Stripe.Subscription = data.object;
+        const customerId = subscription.customer as string;
+        const customer = (await this.stripe.customers.retrieve(
+            customerId,
+        )) as StripeCustomer;
+        const { current_period_start, current_period_end } = subscription;
+        const { name, email } = customer;
+        const emailPayload = generateSubscriptionEmailPayload({
+            email,
+            subject: 'Subscription',
```
Comment (MaxPoplavskiy) [inline backend/src/bundles/stripe-events/handlers/subscription-created-event-handler.ts:37]: consider using enum

**Review by MaxPoplavskiy — CHANGES_REQUESTED** [review]

File: backend/src/bundles/stripe-events/helpers/subscription-email-generator.ts [hunk 2]
```diff
@@ -0,0 +1,47 @@
+import { type SendMailOptions } from 'nodemailer';
+
+import { EmailPayload } from '../constants/email-payload.js';
+import { type SubscriptionDetails } from '../types/types';
+
+const loadAndReplaceTemplate = (variables: Record<string, string>): string => {
+    let htmlContent: string = EmailPayload.HTML;
```
Comment (Icerock10) [inline backend/src/bundles/stripe-events/helpers/subscription-email-generator.ts:7]: no point to declare it as string

File: backend/src/bundles/stripe-events/helpers/subscription-email-generator.ts [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+    email,
+    subject = 'Subscription',
+    name,
+    current_period_start,
+    current_period_end,
+}: SubscriptionDetails): SendMailOptions => {
+    const startDate = new Date(
+        current_period_start * 1000,
+    ).toLocaleDateString();
+    const endDate = new Date(current_period_end * 1000).toLocaleDateString();
```
Comment (Icerock10) [inline backend/src/bundles/stripe-events/helpers/subscription-email-generator.ts:26]: What is the magic number 1000?
maybe separate function formatDate for example to follow DRY?

**Review by Icerock10 — CHANGES_REQUESTED** [review]

File: backend/src/bundles/stripe-events/handlers/subscription-created-event-handler.ts (hunk 1 above)
Comment (jurok3x) [reply backend/src/bundles/stripe-events/handlers/subscription-created-event-handler.ts:37]: done

File: backend/src/bundles/stripe-events/helpers/subscription-email-generator.ts (hunk 2 above)
Comment (jurok3x) [reply backend/src/bundles/stripe-events/helpers/subscription-email-generator.ts]: done

File: backend/src/bundles/stripe-events/helpers/subscription-email-generator.ts (hunk 3 above)
Comment (jurok3x) [reply backend/src/bundles/stripe-events/helpers/subscription-email-generator.ts]: done

**Review by Icerock10 — APPROVED** [review]

File: backend/src/bundles/stripe-events/types/subscription-details.ts [hunk 4]
```diff
@@ -0,0 +1,11 @@
+type SubscriptionDetails = {
+    email: string;
+    subject: string;
+    name: string;
+    current_period_start: number;
+    current_period_end: number;
+    items?: { description: string }[];
```
Comment (MaxPoplavskiy) [inline backend/src/bundles/stripe-events/types/subscription-details.ts:7]: consider moving to a separate type

**Review by MaxPoplavskiy — CHANGES_REQUESTED** [review]

File: backend/src/bundles/stripe-events/helpers/subscription-email-generator.ts [hunk 5]
```diff
@@ -0,0 +1,57 @@
...
+
+const loadAndReplaceTemplate = (variables: Record<string, string>): string => {
+    let htmlContent = EmailPayload.HTML;
+    for (const key of Object.keys(variables)) {
+        const regex = new RegExp(`{${key}}`, 'g');
+        htmlContent = htmlContent.replace(regex, variables[key]);
+    }
+
+    return htmlContent;
+};
```
Comment (Icerock10) [inline backend/src/bundles/stripe-events/helpers/subscription-email-generator.ts:17]: I would move helper functions to a separate file

**Review by Icerock10 — CHANGES_REQUESTED** [review]

File: backend/src/bundles/stripe-events/helpers/subscription-email-generator.ts [hunk 6]
```diff
@@ -0,0 +1,57 @@
...
+
+const formatDate = (timeStamp: number): string => {
+    const date = new Date(timeStamp * MILLIS_TO_SECOND);
+    const day = date.getDate().toString().padStart(2, '0');
+    const month = (date.getMonth() + 1).toString().padStart(2, '0');
+    const year = date.getFullYear();
+
+    return `${day}/${month}/${year}`;
+};
```
Comment (Icerock10) [inline backend/src/bundles/stripe-events/helpers/subscription-email-generator.ts:26]: I would move the helper functions to a separate file

File: backend/src/bundles/stripe-events/helpers/subscription-email-generator.ts (hunk 5 above)
Comment (Icerock10) [inline backend/src/bundles/stripe-events/helpers/subscription-email-generator.ts:17]:
```suggestion
const loadAndReplaceTemplate = (variables: Record<string, string>): string => {
    let htmlContent = EmailPayload.HTML;
    for (const key of Object.keys(variables)) {
        const regex = new RegExp(`{${key}}`, 'g');
        htmlContent = htmlContent.replace(regex, variables[key]);
    }

    return htmlContent;
};
```

```suggestion
  let htmlContent = EmailPayload.HTML;
    for (const [key, value] of Object.entries(variables)) {
        htmlContent = htmlContent.split(`{${key}}`).join(value);
    }
    return htmlContent;
```

regex can be avoided here

**Review by Icerock10 — CHANGES_REQUESTED** [review]

File: backend/src/bundles/stripe-events/helpers/subscription-email-generator.ts (hunk 6 above)
Comment (Icerock10) [reply backend/src/bundles/stripe-events/helpers/subscription-email-generator.ts:26]: .

**Review by Icerock10 — CHANGES_REQUESTED** [review]

**Comment by vlad-blip** [comment]
[create StripeEventDispatcher.](https://trello.com/c/j8bp6eCC/146-create-stripeeventdispatcher)

**Review by MaxPoplavskiy — APPROVED** [review]

**Review by honkster-nazarii — APPROVED** [review]

**Review by RevolveStrm — APPROVED** [review]

**Review by Vitaliy-Kalenichenko — APPROVED** [review]

**Pull request merged**